### ISSUE TITLE
[Feature Request] Generate groups with a column in DataFrame

### ISSUE THREAD

**Issue by kescobo** [JuliaPlots/StatPlots.jl#44]
It would be great to be able to take subsets from a DataFrame to plot different lines or bar groups. So for example, if I have:

```julia
df = DataFrame(a=[1,2,3,4], b=["foo","bar","foo","bar"])
```

I'd like to generate a plot like so:

```julia
using StatPlots

groupedbar(df, :a, group=:b)
```

and get:

<img width="483" alt="screen shot 2017-03-05 at 9 54 48 am" src="https://cloud.githubusercontent.com/assets/3502975/23588920/925b05ae-0192-11e7-8c09-a37c72940ae8.png">

(this was generated by doing `groupedbar([[1,3] [2,4]], lab=["foo", "bar"])`)

**Comment by mkborregaard** [comment]
https://github.com/JuliaPlots/StatPlots.jl/pull/45

**Comment by mkborregaard** [comment]
I just copied in the discussion from gitter to keep it here:

@mkborregaard:
I think it is because groupedbar is a user recipe and works on x and y, so 1) the DataFrames recipe never gets called, and 2) the extractGroupArgs for DataFrames isn't called. Making it a series recipe wouldn't be straigthforward either, because groups creates different series, but all of the series need to know the x or y values of the other series to plot correctly. I have tried to work out how to get around this on several occasions

I'd think a f(g::GroupedBar, df::DataFrame) recipe would solve issue 1 above, but the groups thing is a bit trickier

mar 05 21:04
... I was rather hoping you'd say something like "you're making it more difficult than it is – you just have to trust in the magic of the recipe" :grinning:

12:50
There is a WIP PR to fix this here JuliaPlots/StatPlots.jl#45 - it solves the DataFrame issue, but there is another issue, that groups don't behave well for groupedbar, as the grouping happens before the user recipe is called - it thus calls the user recipe twice on a vector rather than once on an array. This creates results that look like this:

Also tbreloff I was surprised that to solve it I had to do GroupedBar(args...) = GroupedBar(args) https://github.com/JuliaPlots/StatPlots.jl/pull/45/files#diff-f33c65d965119ad9947498ccafec764aR2 to get the args right, should that be a general formula for userplot or would that have some unintended consequences?


@kescobo
14:23
mkborregaard Thanks for taking a stab at this. Another thing I encountered when trying to solve this on my own is that there might be times when you want the bars to be in a different order than they are in the dataframe. For example, I have some data that's labeled "first", "second", "penultimate", "last" in the df, and I want it plotted in that order, but the df is sorted alphabetically (so "first", "last", "penultimate", "second"). Would be nice to be able to pass an argument to get the correct order but I'm not sure how much that complicates things.
When I plotted this data like a year ago with gadfly, there was a way to do it, but I don't remember the specifics. I just ended up writing a function that I can use with sort!() to get the order correct for plots

@mkborregaard
14:49
If you use a Matrix like in your example you should be able to just supply the x values as the first arg to groupedbar. Again the groups are tricky

15:13
I think perhaps a more general solution would be to give up on passing a matrix or the group argument and use the same format as boxplots and make groupedbar a series recipe.
eg. plot(df, :b, :a, st = groupedbar) or groupedbar(df, :b, :a). I think that is a better general solution than the open PR.
But would be breaking.

**Comment by mkborregaard** [comment]
But @kescobo say this format worked, or the `groups` that you are requesting. That would specify the colour of the bars. But what would specify in which group each bar of that colour would go (e.g. where on the x axis? It seems to me you would need both a `group` and an x axis value. In which case my seriestype idea is off the table again.
And your syntax would have to be:
```julia
using DataFrames, StatPlots
df = DataFrame(a=[1,2,3,4], b=["foo","bar","foo","bar"], c= [1,1,2,2])
groupedbar(df, :c, :a, group=:b)
```
right?

**Comment by mkborregaard** [comment]
So the conclusion is that we need a keyword and cannot use the existing `group` keyword. I will implement a version with a new keyword, e.g. `bargroups`?

**Comment by kescobo** [comment]
> But what would specify in which group each bar of that colour would go (e.g. where on the x axis - the order?

Just to be certain that we're referring to the same thing, when I say "group" in this context, I'm referring to bars that are close together (multiple colors). I'm not sure what word I'd use for bars being the same color, but maybe something like "condition."

Let me use an actual example I have (you can see it [here](https://nbviewer.jupyter.org/github/kescobo/gender-comp-bio/blob/texinit/notebooks/gender_stats.ipynb)). I have a table that looks like this:

|   |Dataset|Position       |Mean	            |Lower	            |Upper              |
|---|-------|---------------|-------------------|-------------------|-------------------|
|1	|bio	|first	        |0.37554899967549205|0.37341618040791885|0.37760934519023026|
|2	|bio	|last	        |0.2445807802726165	|0.2425217412944195	|0.24667229687530196|
|3	|bio	|other	        |0.36811105488635143|0.36672005662522206|0.36955886435787955|
|4	|bio	|penultimate	|0.2794508567134617	|0.2769456272599794	|0.28192603507064085|
|5	|bio	|second	        |0.37867720975402264|0.376309021032719	|0.38124814138903007|
|6	|comp	|first	        |0.31628457043398167|0.31216277156822253|0.32071948984831855|
|7	|comp	|last	        |0.20728100188266085|0.2035118072511263	|0.21131039035313787|
|8	|comp	|other	        |0.33066245511496395|0.3278958261091234	|0.3334795748392778 |
|9	|comp	|penultimate	|0.23616028172389428|0.2314216315959131	|0.24072637668427668|
|10	|comp	|second	        |0.3219207849916515	|0.3173954530484164	|0.3265934968596715 |

What I want is a bar graph with one color for "bio", one color for "comp" (so color based on the `:Dataset` column), and the bar groups based on `:Position`, but I'd like to be able to specify the order of the groups. 

So something that would be intuitive would be something like:

```
groupedbar(df, :Mean, 
        groups=(:Position, ["first", "second", "other", "penultimate","last"]), 
        color=:Dataset)
```

**Comment by mkborregaard** [comment]
Hm, not simple at all. I wonder if this could be generalized so it would be intuitive for all (of course we cannot reuse the regular keywords to plot).
Currently this would be 
```julia
df2 = unstack(df, :Position, :Dataset, :Mean)
df2 = df2[[1,5,3,4,2],:]
groupedbar(df2[:Position], Array(df2[:,2:3]))
```
The question is whether an automagic version with keywords would really be much more intuitive?

**Comment by kescobo** [comment]
>The question is whether an automagic version with keywords would really be much more intuitive?

Perhaps not. Maybe the best solution would just be to have a few different examples with solutions documented. FWIW - it would not have occurred to me to use unstack for example. 

That said, though I could be wrong, it seems like this is not a terribly uncommon need. Perhaps sorting is a bridge too far in terms of complexity, but being able to specify which rows belong to which bar group and which color is something I'd do all the time.

**Comment by kescobo** [comment]
Generic function to generate sort order:

```julia
function getorder(v, o)
    order = Dict(key => ix for (ix, key) in enumerate(o))
    return([order[k] for k in v])
end

getorder(df[:Position], ["first", "second", "other", "penultimate", "last"])
```
returns: [1,1,5,5,3,3,4,4,2,2]

**Comment by piever** [comment]
Closed by [#1062](https://github.com/JuliaPlots/Plots.jl/pull/1062) and #83
